### DISCUSSION TITLE
Override/Extend default register mutation

### DISCUSSION THREAD

**Discussion by dominikklein** [graphql-devise/graphql_devise#219] [Q&A]
I tried to extend the default register mutation.

For this I added a own register mutation like this:

```
module Mutations
  class Register < GraphqlDevise::Mutations::Register
    argument :name, String, required: true
  end
end
```

And added a registration for this mutation in the `mutation_type.rb`:

`field :user_register, mutation: Mutations::Register`

But i have the following error:

```
{
	"data": {
		"userRegister": null
	},
	"errors": [
		{
			"message": "userRegister field requires authentication",
			"locations": [
				{
					"line": 2,
					"column": 2
				}
			],
			"path": [
				"userRegister"
			],
			"extensions": {
				"code": "AUTHENTICATION_ERROR"
			}
		}
	]
}
```

I also tried this: `field :user_register, mutation: Mutations::Register, authenticate: false`, which leads to a different error:

https://gist.github.com/dominikklein/b559ef9037d450fba33028427821c5a6

Can maybe someone help me, what I'm doing wrong?

**Comment by 00dav00** [comment]
Hi @dominikklein , I think that for this particular mutation you need to call `super`, like in the [specs example](https://github.com/graphql-devise/graphql_devise/blob/master/spec/dummy/app/graphql/mutations/register.rb). I can check with more detail later on but I would recommend checking this option in the mean time. 

BWT `authenticate: false` should be correct, since the user won't have a way to create a token before registering.

**Reply by dominikklein** [reply]
Hm, i added now the same code like in the spec example, but same error with `authenticate: false`:a

```
module Mutations
  class Register < GraphqlDevise::Mutations::Register
    argument :name, String, required: true

    field :user, Types::UserType, null: true

    def resolve(email:, **attrs)
      original_payload = super
      original_payload.merge(user: original_payload[:authenticatable])
    end
  end
end
```

**Reply by 00dav00** [reply]
Hi @dominikklein , reading your backtrace more carefully it looks like the gem is not abel to find the resource you try to use in that mutation. Have you there these sections of these readme ([1](https://github.com/graphql-devise/graphql_devise#mounting-auth-schema-on-a-separate-route), [2](https://github.com/graphql-devise/graphql_devise#mounting-operations-in-an-existing-schema))?

Could you share the way you are mounting the resources?

**Reply by mcelicalderon** [reply]
@dominikklein I think you are adding your mutation directly to your `MutationType`, right? Perhaps you should add your custom mutation to the resource mounter in the operations key https://github.com/graphql-devise/graphql_devise#available-mount-options

```ruby
class DummySchema < GraphQL::Schema
  # It's important that this line goes before setting the query and mutation type on your
  # schema in graphql versions < 1.10.0
  use GraphqlDevise::SchemaPlugin.new(
    query:            Types::QueryType,
    mutation:         Types::MutationType,
    resource_loaders: [
      GraphqlDevise::ResourceLoader.new(User, operations: { register: Mutations::Register })
    ]
  )

  mutation(Types::MutationType)
  query(Types::QueryType)
end
```
The same can be done if mounting in a separate route (provide `operations` to the mount method). That does `authenticate: false` internally.

Adding it directly to your mutation type should also work, but not sure how you are using the gem. You should exclude the gem's register mutation when mounting if the field for your mutation has the same name

**Reply by dominikklein** [reply]
I use this:

```
  use GraphqlDevise::SchemaPlugin.new(
    query: Types::QueryType,
    mutation: Types::MutationType,
    resource_loaders: [
      GraphqlDevise::ResourceLoader.new(User, only: %i[login logout])
    ]
  )

  mutation(Types::MutationType)
  query(Types::QueryType)
  ```
  
  And the registration of my own mutation is inside of the mutation type, but tested also the way over the operations-key, both are not working with the same error.

**Reply by mcelicalderon** [reply]
@dominikklein I can't think of anything else that might be wrong at this point. I think the only way we could help you is if you are able to push a test project where you can replicate this problem
